### ISSUE TITLE
Considering of negative samples in total loss processing  

### ISSUE THREAD

**Issue by ARI777** [pierluigiferrari/ssd_keras#69]
Hi, 
I cannot understand, why the total loss processing in "keras_ssd_loss.py" is “normalized” on the number of positive samples only? 

Line 201:  total_loss = (class_loss + self.alpha * loc_loss) / tf.maximum(1.0, n_positive)

In my understanding it is valid for localization only. In the case of classification it is not exactly correct and number of negatives should be used also, like:

total_loss = class_loss / (tf.to_float(n_negative_keep) + tf.maximum(1.0, n_positive)) +  (self.alpha * loc_loss) / tf.maximum(1.0, n_positive)

In practice, by constant neg_pos_ratio and correct weighting by self.alpha, the current implementation will work well also, I suppose… But if you have no positives in a batch?

Thank you for your comments!
Best regards
Andrej

**Comment by pierluigiferrari** [comment]
There are a couple of aspects to this:

1. Normalizing by the number of matched boxes is how the authors of the original implementation designed their loss function. Since the goal of this project was to create a faithful port of the original Caffe implementation, the question of whether or not this particular design choice is ideal doesn't really arise.
2. There is no one correct way to normalize a loss function. This is just one sensible way to do it.
3. `n_negative_keep` is directly proportional to `n_positive` via the `neg_pos_ratio` (as you noted correctly). The only point of normalizing the loss is to make its magnitude independent of the number of objects in the batch. Hence, as long as you divide the loss by any number that is proportional to the number of objects in the batch, this requirement is fulfilled.
